File: bin/kival/src/commands/objects/events.rs
//! Object event-listing command.

use argx::{Args, argx};
use kival_cli::runner::CliContext;
use kival_sdk::{Event, EventListParams, EventOrder, ListResponse};
use uuid::Uuid;

use super::{ObjectCommandError, ObjectTargetArgs, object_error_codes};
use crate::{
    commands::events::print_event_line,
    utils::{
        args::DEFAULT_LIST_LIMIT,
        credentials::authenticated_client,
        output::{OutputMode, print_empty_list, print_output},
    },
};

object_error_codes! {
    pub(crate) enum ObjectEventsErrorCode {
        failures {
            AuthenticationRequired => ("authentication.required", AuthenticationRequired),
            PermissionDenied => ("permission.denied", PermissionDenied),
            InvalidArgument => ("invalid.argument", InvalidArgument),
            ResourceNotFound => ("resource.not_found", ResourceNotFound),
            ServerUnavailable => ("server.unavailable", ServerUnavailable),
            RateLimited => ("rate_limited", RateLimited),
            RequestFailed => ("request.failed", RequestFailed),
            InvalidField => ("output.invalid_field", InvalidField),
            InvalidProjection => ("output.invalid_projection", InvalidProjection),
            InvalidCursor => ("invalid.cursor", InvalidCursor),
            Internal => ("internal", Internal),
        }
        objects { ObjectNotFound => ("object.not_found", NotFound) }
    }
}

/// Error returned by the corresponding command handler.
type ObjectEventsError = ObjectCommandError<ObjectEventsErrorCode>;

/// Arguments for `kival objects events`.
#[derive(Debug, Args)]
pub struct ObjectEventsCommand {
    /// Object target.
    #[argx(flatten)]
    pub target: ObjectTargetArgs,
    /// Maximum number of events to return.
    #[argx(long, default = DEFAULT_LIST_LIMIT)]
    pub limit: Option<i64>,
    /// Return events with a global sequence number strictly greater than SEQUENCE.
    #[argx(long)]
    pub after_sequence: Option<i64>,
    /// Return events with a global sequence number strictly less than SEQUENCE.
    #[argx(long)]
    pub before_sequence: Option<i64>,
    /// Event sequence ordering.
    #[argx(long, default = EventOrder::Asc)]
    pub order: EventOrder,
    /// Filter by event kind.
    #[argx(long)]
    pub event_kind: Option<String>,
    /// Filter by actor user ID.
    #[argx(long)]
    pub actor_user_id: Option<Uuid>,
    /// Filter by target user ID.
    #[argx(long)]
    pub target_user_id: Option<Uuid>,
    /// Filter by group ID.
    #[argx(long)]
    pub group_id: Option<Uuid>,
}

#[argx(handler = run)]
impl ObjectEventsCommand {
    /// Run `kival objects events`.
    ///
    /// # Errors
    ///
    /// Returns an error if events cannot be listed.
    pub(crate) async fn run(
        self,
        ctx: CliContext,
        output: OutputMode,
    ) -> Result<ListResponse<Event>, ObjectEventsError> {
        let params = EventListParams {
            limit: self.limit,
            after_sequence: self.after_sequence,
            before_sequence: self.before_sequence,
            order: self.order,
            event_kind: self.event_kind,
            actor_user_id: self.actor_user_id,
            target_user_id: self.target_user_id,
            object_id: Some(self.target.object_id),
            group_id: self.group_id,
        };
        let client = authenticated_client(&ctx)?;
        let response = client
            .list_object_events(self.target.workspace_id, self.target.object_id, &params)
            .await?;
        print_output(&output, &response, || {
            if response.items.is_empty() {
                print_empty_list("events");
            } else {
                for event in &response.items {
                    print_event_line(event);
                }
            }
        })?;
        Ok(response)
    }
}

File: sdk/typescript/src/actions/events.ts
import { pathId, requestInit, withParams } from "../internal/utils.js";
import type { Event, EventListParams, ListResponse, UUID } from "../types.js";
import { setNumber, setString } from "./params.js";
import type { KivalClientBase, WithSignal } from "./types.js";

/** Parameters for {@link listObjectEvents}. */
export type ListObjectEventsParameters = WithSignal<
  EventListParams & {
    workspaceId: UUID;
    objectId: UUID;
  }
>;

/** Return type for {@link listObjectEvents}. */
export type ListObjectEventsReturnType = ListResponse<Event>;

/** Lists events for an object. */
export function listObjectEvents(
  client: KivalClientBase,
  parameters: ListObjectEventsParameters,
): Promise<ListObjectEventsReturnType> {
  const { workspaceId, objectId, ...options } = parameters;
  return client.requestJson<ListObjectEventsReturnType>(
    withParams(
      `/workspaces/${pathId(workspaceId)}/objects/${pathId(objectId)}/events`,
      eventParams(options),
    ),
    requestInit({}, options.signal),
  );
}

/** Parameters for {@link listWorkspaceEvents}. */
export type ListWorkspaceEventsParameters = WithSignal<
  EventListParams & {
    workspaceId: UUID;
  }
>;

/** Return type for {@link listWorkspaceEvents}. */
export type ListWorkspaceEventsReturnType = ListResponse<Event>;

/** Lists events in a workspace. */
export function listWorkspaceEvents(
  client: KivalClientBase,
  parameters: ListWorkspaceEventsParameters,
): Promise<ListWorkspaceEventsReturnType> {
  const { workspaceId, ...options } = parameters;
  return client.requestJson<ListWorkspaceEventsReturnType>(
    withParams(`/workspaces/${pathId(workspaceId)}/events`, eventParams(options)),
    requestInit({}, options.signal),
  );
}

/** Parameters for {@link listEvents}. */
export type ListEventsParameters = WithSignal<EventListParams>;

/** Return type for {@link listEvents}. */
export type ListEventsReturnType = ListResponse<Event>;

/** Lists global events. */
export function listEvents(
  client: KivalClientBase,
  parameters: ListEventsParameters = {},
): Promise<ListEventsReturnType> {
  return client.requestJson<ListEventsReturnType>(
    withParams("/events", eventParams(parameters)),
    requestInit({}, parameters.signal),
  );
}

function eventParams(options: WithSignal<EventListParams>) {
  const params = new URLSearchParams();
  setNumber(params, "limit", options.limit);
  setNumber(params, "after_sequence", options.after_sequence);
  setNumber(params, "before_sequence", options.before_sequence);
  if (options.order === "desc") params.set("order", "desc");
  setString(params, "event_kind", options.event_kind);
  setString(params, "actor_user_id", options.actor_user_id);
  setString(params, "target_user_id", options.target_user_id);
  setString(params, "object_id", options.object_id);
  setString(params, "group_id", options.group_id);
  return params;
}
